### ISSUE TITLE
beg for experience of setting parameters

### ISSUE THREAD

**Issue by phoenixbai** [samsungsds-rnd/deepspeech.mxnet#18]
Hi,
I am trying to feed deepspeech with my own dataset, but now sure how to set certain parameters like width, height, channel etc.
Could you help? Also, could you share the experience with setting the training parameters as well?
Thanks in advance!

**Comment by Soonhwan-Kwon** [comment]
It depends on sampling rate of your data source. If it is 8kHz,  width should be 81(when it is window size of 20ms and it is defined at stt_datagenerator.py) (and also we assumed height and channel will be 1). 
I recommend you to be careful with buckets arguments, because it consumes more cpu memory and takes more time when it is switching context of the bucket size with many buckets(the program can be end up core dump because of too much memory usage). 
So first try with one bucket size(maximum size of your input audio length) and check it met with the gpu memory, then try with more buckets.
If you have more question feel free to contact with me. Thank you.

**Comment by phoenixbai** [comment]
@Soonhwan-Kwon 
hi, I am training with 10000 audio files with wav format, and using with default parameters as below:
```
batch_size = 10
max_duration = 17.0
language = en
width = 161
channel_num = 32
conv_layer1_filter_dim = [11, 41]
conv_layer1_stride = [2, 2]
conv_layer2_filter_dim = [11, 21]
conv_layer2_stride = [1, 2]

num_rnn_layer = 1
num_hidden_rnn_list = [600]
num_hidden_proj = 0

num_rear_fc_layers = 0
num_hidden_rear_fc_list = []
act_type_rear_fc_list = []
rnn_type = gru
num_epoch = 50
learning_rate = 0.003
num_epoch = 50
learning_rate = 0.003
# constant learning rate annealing by factor
learning_rate_annealing = 1.1
initializer = Xavier
init_scale = 2
factor_type = in
# show progress every nth batches
show_every = 10
save_optimizer_states = True
normalize_target_k = 5000
optimizer = adam
optimizer_params_dictionary={"beta1":0.9,"beta2":0.999}
clip_gradient = 0
weight_decay = 0.
```

after trained around 14 hours, it starts to predict nonsense like below:

> [    INFO][2018/05/24 10:59:28.293] label: tu2 wei2 xian4 zheng4 xie2 fu4 zhu3 xi2 gao1 ji2 nong2 yi4 shi1 mi4 xiu4 yun2 zai4 sun1 lou2 xiang1 guo1 lou2 cun1 gei3 mian2 nong2 jiang2 jie3 liang2 zhong3 de5 shi3 yong4 fang1 fa3
[    INFO][2018/05/24 10:59:28.293] pred : ta1 e4 yi2 xia4 ziai4 xia4 ce4 zhuxy4 bai4 jie e2 yi s4 y yi4 xio2 li2 zai4 shen4 deu4 xia g le2 chan2 yen4 yia4 le2 jian2 ji yan2 zu1 ze5 shi4 yen2 fan4 fa yen , cer: 0.472527 (distance: 86/ label length: 182)
[    INFO][2018/05/24 10:59:28.387] label: ta1 dai4 biao3 e2 yu3 xue2 yuan4 xiang4 nai4 na4 fu1 ren5 zeng4 song4 le5 yi1 tao4 yu3 pu3 xi1 jin1 e2 yu3 xue2 yuan4 he2 bian1 de5 jiao4 cai2 dong1 fang1 e2 yu3
[    INFO][2018/05/24 10:59:28.387] pred : ta4 gai4 dan huei4y4 xue2 yan4 qian4 nai4 da4 f le2 zan4 qu4 e2 lan2  yuan4 zeu1 xi4 jie2 lueyi4 x ya4 fe4 hao4 be4 de5 zhian1 cen den4 fao2 huei2 la , cer: 0.527950 (distance: 85/ label length: 161)
[    INFO][2018/05/24 10:59:28.480] label: ta1 niang2 jia5 xiong1 di4 zi3 mei4 shi2 yi2 ge4 ta1 lao3 da4 cong2 xiao3 jiu4 gei3 jiu4 jia1 zuo4 yang2 nv3 gai3 ming2 lv3 ai4 zhi4 min3 nan2 hua4 shi4 ai4 di4
[    INFO][2018/05/24 10:59:28.480] pred : ta4 zian la4 qeng2 nin4 tn2 fe qen2 yi2 ge4 a2la2 ga4 zan2 xian2 da4 bin4 zen4 jian4 zen4 yang yie2 bai4 lin2 le2 ya2 zi2 nin4 ya2 wa2 ci2 dai2 gi ei , cer: 0.493750 (distance: 79/ label length: 160)
[    INFO][2018/05/24 11:00:08.310] label: tian1 gong1 bu4 zuo4 mei3 dang4 tian1 wan3 kai1 shi3 zhuan3 yin1 xia4 yu3 ci4 ri4 qing1 chen2 reng2 yin1 yun2 mi4 bu4 ren2 men5 ju3 sang4 de5 zhun3 bei4 che4 li2
**[    INFO][2018/05/24 11:00:08.311] pred :  , cer: 1.000000 (distance: 161/ label length: 161)
[    INFO][2018/05/24 11:00:08.315] label: yang2 zhuo2 yong1 hu2 chou1 shui3 xu4 neng2 dian4 chang3 shi4 mu4 qian2 shi4 jie4 shang5 hai3 ba2 zui4 gao1 de5 chou1 shui3 xu4 neng2 dian4 chang3 ye3 shi4 xi1 zang4 zui4 da4 de5 neng2 yuan2 ji1 di4
[    INFO][2018/05/24 11:00:08.315] pred :  , cer: 1.000000 (distance: 198/ label length: 198)
[    INFO][2018/05/24 11:00:08.320] label: jiang1 xi1 hu2 kou3 xian4 shuang1 zhong1 shui2 chan3 chang3 yu2 min2 zhang1 hou4 yu2 ri4 qian2 jiang1 yi1 tiao2 bu3 huo4 ru4 wang3 de5 xing2 si4 bian1 yu2 de5 yan1 zhi5 yu2 fang4 hui2 jiang1 zhong1
[    INFO][2018/05/24 11:00:08.320] pred :  , cer: 1.000000 (distance: 197/ label length: 197)
[    INFO][2018/05/24 11:00:08.357] label: lao3 di4 xiong5 lia3 ting1 shuo1 hai3 min2 yao4 wa1 chi2 yang3 yu2 jiu4 cou4 dao4 yi2 kuai4 la1 tan2 guo4 kan4 neng2 bu5 neng2 zai4 hai3 min2 zhe4 li3 ru4 ge4 gu3
[    INFO][2018/05/24 11:00:08.357] pred :  , cer: 1.000000 (distance: 162/ label length: 162)
[    INFO][2018/05/24 11:00:08.361] label: ta1 niang2 jia5 xiong1 di4 zi3 mei4 shi2 yi2 ge4 ta1 lao3 da4 cong2 xiao3 jiu4 gei3 jiu4 jia1 zuo4 yang2 nv3 gai3 ming2 lv3 ai4 zhi4 min3 nan2 hua4 shi4 ai4 di4
[    INFO][2018/05/24 11:00:08.361] pred :  , cer: 1.000000 (distance: 160/ label length: 160)
[    INFO][2018/05/24 11:00:08.391] label: tan4 suo3 wei2 rao4 jing1 ji4 jian4 she4 zhua1 dang3 jian4 de5 you3 xiao4 tu2 jing4 luo2 tian2 nong2 cun1 dang3 zhi1 bu4 kai1 zhan3 chuang4 jian4 xiao3 kang1 cun1 jing4 sai4
[    INFO][2018/05/24 11:00:08.391] pred :  , cer: 1.000000 (distance: 173/ label length: 173)
[    INFO][2018/05/24 11:00:08.447] label: ta1 shi4 li4 yong4 guang1 bo1 sao3 miao2 pei4 he2 ji1 guang1 guo4 lv4 yi3 hua4 xue2 wu4 xi1 qu4 zao4 mo2 zui4 hou4 yi3 yi1 zhong3 dian4 nao3 ji4 shu4 zhi4 zuo4 er2 cheng2
[    INFO][2018/05/24 11:00:08.448] pred :  , cer: 1.000000 (distance: 170/ label length: 170)
[    INFO][2018/05/24 11:00:08.452] label: shuo1 na4 shi2 hou5 wei1 e2 da4 bie2 shan1 meng3 lie4 zhen4 chan4 le5 yi2 xia4 kuang2 feng1 zhou4 qi3 lve4 guo4 shan1 ye3 mang2 mang2 hei1 ye4 yi1 pian4 ku1 sheng1
[    INFO][2018/05/24 11:00:08.452] pred :  , cer: 1.000000 (distance: 163/ label length: 163)
[    INFO][2018/05/24 11:00:08.481] label: ling4 wai4 kuang4 qu1 mei2 tan4 guan2 li3 ju2 dui4 xiang1 zhen4 kuang4 de5 kai1 cai3 fan4 wei2 zuo4 le5 ming2 que4 hua4 fen1 yan2 jin4 yue4 jie4 yue4 ceng2 kai1 cai3
[    INFO][2018/05/24 11:00:08.481] pred :  , cer: 1.000000 (distance: 165/ label length: 165)
[    INFO][2018/05/24 11:00:08.486] label: zhe4 shi4 quan2 sheng3 ge4 ji2 nong2 zi1 gong1 si1 zhuan1 men2 xia4 pai4 de5 zhi1 nong2 song4 fei2 che1 song4 huo4 xia4 xiang1 bang1 zhu4 nong2 min2 chu3 bei4 chun1 geng1 yong4 fei2**

Is it due to gradient explosion? I tried with clip_gradient=100 (set in random), not helpful at all.
Tried with sgd as well, worse.
could you help? is my network structure appropriate for the task? what is the better way for optimizing it? 

p.s. my dataset is Mandarin and label is spellings which consists of a-z, 0-5.

thanks

**Comment by Soonhwan-Kwon** [comment]
The main cause is that the architecture you used needs more capacity. You used one layer of rnn and it is suitable for toy example so you need to stack more rnn layers for example
num_rnn_layer=3
num_hidden_rnn_list=[600, 600,600]
and if it is not enough then try num_rnn_layer=5,7 with more hidden unit.

**Comment by Soonhwan-Kwon** [comment]
And I'm afraid that 0-5 may harm your accuracy of the model. I think it is more suitable to deal with numbers in post processing or language model because they will conflict with your alphabet labels and also have very different seq length compare to the alphabet.
